node_modules_koding/bongo/lib/bongo.coffee

```diff
@@ -105,71 +120,123 @@ module.exports = class Bongo extends EventEmitter
+
+    model.on 'needsPopulated', (def, _path) =>
+
+      _path = if Array.isArray _path then _path else _path.split '.'
+      names = JsPath.getAt def, _path
+
       unless Array.isArray names
         names = [names]
-        JsPath.setAt def, path, names
-      names.forEach (name, i)=>
+        JsPath.setAt def, _path, names
+
+      names.forEach (name, i) =>
         if target = @models[name]
-          JsPath.setAt def, path.concat(i), target
-      @on 'newModel', (newName, newModel)->
+          JsPath.setAt def, _path.concat(i), target
+
+      @on 'newModel', (newName, newModel) ->
         if newName in names
-          JsPath.setAt def, path.concat(names.indexOf newName), newModel
+          JsPath.setAt def, _path.concat(names.indexOf newName), newModel
+
+    return name
+
+
+  registerModel: (path) ->
+
+    cons = require path
+
+    if cons.name? and ('function' is typeof cons)
+      @defineModel cons.name, cons
+
+    else
+      models = []
+      for own name, konstructor of cons
+        models.push @defineModel name, konstructor
+
+      models
+
 
   findCode = (paths) ->
-    Promise.map paths, (path) ->
-      fs.statAsync(path).then (stats) ->
+    Promise.map paths, (_path) ->
+      fs.lstatAsync(_path).then (stats) ->
         # in the case that this is a directory, recurse into it
         if stats.isDirectory()
-          fs.readdirAsync(path).map (dir) ->
-            nodePath.join path, dir
+          fs.readdirAsync(_path).map (dir) ->
+            path.join _path, dir
           .then findCode
         # otherwise, just collect the path
-        else [path]
+        else [_path]
     # flatten the result
     .reduce((a, b) -> a.concat b)
     # if coffee or js file include it
-    .filter (path) -> return /(\.coffee|\.js)$/.test(path)
+    .filter (_path) -> return /(\.coffee|\.js)$/.test(_path)
     # if test file return false and exclude file
-    .filter (path) -> return !isTestFile path
+    .filter (_path) -> return !(/\.test\.coffee$/.test _path)
 
 
-  endsWith = (str, suffix) ->
+  loadModels: ->
 
-    return str.toLowerCase().indexOf(suffix.toLowerCase(), str.length - suffix.length) != -1
+    Promise.map @modelIndexPaths, (modelPath, index) =>
 
+      fs.lstatSync modelPath
 
-  isTestFile = (path) ->
+      cache = JSON.parse(fs.readFileSync(modelPath).toString())
 
-    return endsWith path, 'test.coffee'
+      # console.log "Model list found:", modelPath
+      # console.log "Loading #{Object.keys(cache).length} models..."
 
+      Object.keys(cache).map (name) =>
+        # console.log "Loading model:", name
+        @registerModel path.join @modelPaths[index], cache[name]
 
-  registerModel: (cons) ->
-    if cons.name? and ('function' is typeof cons)
-      @defineModel cons.name, cons
-    else
-      @defineModel name, konstructor  for own name, konstructor of cons
+    .catch (err) =>
+      # console.log 'Cache not found or corrupted, re-building...', err.message
```

Comment:
could you fail with exit here if we are in CI ( if CI env var is set )? With that way we will keep the cache up-to-date, in case someone forgets to push `models.json`

Also if we add a descriptive message on what developer should do if this case happens, could help future generations

ie : "if you see this message, please commit models.json as well"